<?xml version="1.0" encoding="UTF-8"?> 
<launch>
    <!-- Parameters -->
    <arg name="use_sim_time"              default="false" />
    <param name="use_sim_time"            value="$(arg use_sim_time)"/>

    <!-- GMSF Input Topics -->
    <arg name="imu_topic_name"            default="/sensors/imu" />
    <arg name="lidar_odometry_topic_name" default="/compslam_lio/odom_aft_mapped_to_init_CORRECTED" />
    <arg name="gnss_topic_name"           default="/rtk_gps_driver/position_receiver_0/ros/navsatfix" />

    <!-- Compslam Topics -->
    <arg name="pointcloud_topic_name"            default="/point_cloud_filter/lidar/point_cloud_filtered" />

    <!-- Servers -->
    <arg name="gnss_coordinates_to_enu_topic_name"  default="/dual_graph_node/get_path_in_enu" />

    <!-- GMSF Config -->
    <arg name="using_gps"                 default="false" />
    <arg name="using_lio"                 default="true" />
    <arg name="imu_gravity_direction"     default="up" />
    <!-- Files -->
    <arg name="graph_config_file"         default="$(find anymal_dual_graph)/config/graph_config.yaml"/>
    <arg name="graph_parameter_file"      default="$(find anymal_dual_graph)/config/graph_params.yaml"/>
    <arg name="extrinsic_file"            default="$(find anymal_dual_graph)/config/extrinsic_params.yaml"/>
    <arg name="gnss_parameter_file"       default="$(find anymal_dual_graph)/config/gnss_params.yaml"/>
    <arg name="trajectory_alignment_parameter_file" default="$(find anymal_dual_graph)/config/traj_align_params.yaml"/>

    <!-- Compslam Config -->
    <arg name="compslam_config_filename"  default="$(find compslam_lio)/config/compslam_lio_VLP16_params.yaml" />

    <!-- Node -->
    <node pkg="anymal_dual_graph" type="anymal_dual_graph_node" name="dual_graph_node" output="screen"> <!--launch-prefix="gdb -ex run -args"-->
        <!-- Launch Parameters -->
        <param name="launch/imuGravityDirection" type="string" value="$(arg imu_gravity_direction)" />
        <param name="launch/usingGnss" type="bool" value="$(arg using_gps)" />
        <param name="launch/usingLio" type="bool" value="$(arg using_lio)" />
        <!-- Parameter files -->
        <rosparam command="load" file="$(arg graph_config_file)" />
        <rosparam command="load" file="$(arg graph_parameter_file)" />
        <rosparam command="load" file="$(arg extrinsic_file)" />
        <rosparam command="load" file="$(arg gnss_parameter_file)" />
        <rosparam command="load" file="$(arg trajectory_alignment_parameter_file)" />
        <!-- Remapping of topics -->
        <remap from="/imu_topic"            to="$(arg imu_topic_name)" />
        <remap from="/lidar_odometry_topic" to="$(arg lidar_odometry_topic_name)" />
        <remap from="/gnss_topic"         to="$(arg gnss_topic_name)" />
        <!-- Remapping of services -->
        <remap from="/gnss_coordinates_to_enu_topic" to="$(arg gnss_coordinates_to_enu_topic_name)" />
    </node>

    <!-- RVIZ for Visualization -->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find anymal_dual_graph)/rviz/gnss_lidar.rviz" />

    <!--node name="gps_path_publisher" type="gps_path_publisher_node" pkg="gps_path_publisher" output="screen" >
        <param name="gpx_filename"                           value="/home/integration/glattpark_anymal_handdrawn.kml" />
        <param name="remove_duplicate_nodes"                 value="true" />
        <param name="output_frame"                           value="enu" />
        <param name="path_creator/auto_orient_nodes"         value="true" />
        <param name="path_creator/reset_node_height"         value="true" />
        <param name="path_creator/keep_relative_node_height" value="true" />
    </node-->

    <!-- TF for Aligning Compslam Map -->
    <node pkg="tf" type="static_transform_publisher" name="compslam_map_to_gmsf_map" args="0 0 0 0 0 0 compslam_lio/camera_init_CORRECTED map_graph_msf 10" />

    <!-- Launch compslam -->
    <include file="$(find compslam_lio)/launch/compslam_lio_asop.launch">
        <arg name="pointcloud_topic_name"           value="$(arg pointcloud_topic_name)" />
        <arg name="imu_topic_name"                  value="$(arg imu_topic_name)" />
        <arg name="config_filename"                 value="$(arg compslam_config_filename)" />
    </include>

    <!-- TF for Aligning MSF -->
    <node pkg="tf" type="static_transform_publisher" name="compslam_map_to_msf_map" args="0 0 0 0 0 0 compslam_lio/camera_init_CORRECTED world 10" />

</launch>
